examples/darwin-framework-tool/commands/common/xpc/XPCServer.mm

```diff
@@ -0,0 +1,256 @@
+/*
+ *   Copyright (c) 2024 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+
+#import "XPCServerRegistry.h"
+
+#import <lib/support/logging/CHIPLogging.h>
+
+@interface XPCServerImpl : NSObject <MTRXPCServerProtocol_MTRDevice>
+@property (nonatomic, strong) id<MTRXPCClientProtocol> clientProxy;
+@property (nonatomic, strong) NSArray<MTRDeviceController *> * controllers;
+@property (nonatomic, strong) dispatch_queue_t callbackQueue;
+
+- (instancetype)init NS_UNAVAILABLE;
++ (instancetype)new NS_UNAVAILABLE;
+- (instancetype)initWithClientProxy:(id<MTRXPCClientProtocol>)proxy controllers:(NSArray<MTRDeviceController *> *)controllers;
+@end
+
+@implementation XPCServerImpl
+- (instancetype)initWithClientProxy:(id<MTRXPCClientProtocol>)proxy controllers:(NSArray<MTRDeviceController *> *)controllers
+{
+    if ([super init]) {
+        _clientProxy = proxy;
+        _controllers = controllers;
+        _callbackQueue = dispatch_queue_create("com.chip.xpc.command", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
+    }
+    return self;
+}
+
+- (MTRDevice *)createDevice:(NSUUID *)uniqueIdentifier nodeID:(NSNumber *)nodeID
+{
+    for (MTRDeviceController * controller in _controllers) {
+        if ([controller.uniqueIdentifier isEqual:uniqueIdentifier]) {
+            return [MTRDevice deviceWithNodeID:nodeID controller:controller];
```

Comment:
Followup: we will want to have these devices held on to, and set delegates on them so they actually get some state in them (like values of attributes), right?